### PR TITLE
Routerstuff

### PR DESCRIPTION
**Pull request by marcelhollerbach**
#### REST
- updated router
- added resources:
  - `AlgorithmResources`
  - `ExperimentResource`
  - `AnswerRatingResource`
#### DB
- updated DB-model & `db.sql`
  - new `Experiments_Platform`-Table
  - rename population -> calibration
  - added missing description to Algorithm_Answer_Quality & Algorithm_Rating_Quality
  - added missing insert-triggers for the algorithm tables
- added new operations-classes: 
  - `AlgorithmOperations`
  - `CalibrationOperations`
  - `TagConstraintsOperations`
- updated most of the other operations-classes
- added new transformation-classes: 
  - `AlgorithmsTransformation`
  - `AnswerRatingTransformation`
  - `CalibrationTransformation`
  - `ExperimentTransformation`
  - `TagConstraintTransformation`
#### Protobuf
- updated the generated protobuf-classes
  - rename population -> calibration
  - added algorithmOption & algorithmList
  - added RatingOptions in Template and Experiment

### CODE REVIEW

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/ExperimentOperations.java [hunk 1]
```diff
@@ -63,4 +68,125 @@ public boolean deleteExperiment(int id) throws IllegalArgumentException {
...
+        if (taskStatuses.isEmpty()) {
+            return Experiment.State.DRAFT;
+        } else if (taskStatuses.contains(TaskStatus.running)) {
+            return Experiment.State.PUBLISHED;
+        } else if (taskStatuses.contains(TaskStatus.stopping)) {
+            return Experiment.State.CREATIVE_STOPPED;
+        } else if (taskStatuses.contains(TaskStatus.stopped)) {
+            return Experiment.State.STOPPED;
+        } else {
+            return Experiment.State.STOPPED; //TODO: finished
```
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/ExperimentOperations.java:102]: I think finished - yes its the only state which is not covered here.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java [hunk 2]
```diff
@@ -54,6 +55,29 @@ protected AbstractOperations(DSLContext create) {
     }
 
     /**
+     * executes the function if the experiment is not running.
+     * @param experimentID the id of the experiment
+     * @param function the function to execute
+     * @param <R> the return type
+     * @return the result of the function
+     */
+    protected <R> R doIfRunning(int experimentID, Function<Configuration, R> function) {
```
Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java]: Docs say **not** running, function is named `doIfRunning`. What's right?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/CalibrationOperations.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+    /**
+     * Returns a range of calibrations starting from {@code cursor}.
+     *
+     * @param cursor Pagination cursor
+     * @param next   {@code true} for next, {@code false} for previous
+     * @param limit  Number of records
+     * @return List of calibrations
+     */
+    public Range<calibration, Integer> getCalibrationsFrom(int cursor, boolean next, int limit) {
```
Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/CalibrationOperations.java]: `calibration` is lowercase, typo?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/transforms/CalibrationTransform.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ */
+public class CalibrationTransform extends AbstractTransform {
+    /**
+     * Converts a calibration record to its protobuf representation.
+     *
+     * @param record calibration record
+     *
+     * @return calibration.
+     */
+    public static calibration toProto(CalibrationRecord record, List<CalibrationAnswerOptionRecord> options) {
```
Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/transforms/CalibrationTransform.java]: `calibration` is again lowercase here, should be `Calibration`.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/Router.java [hunk 5]
```diff
@@ -99,16 +111,26 @@ public void init() {
...
         get("/workers", workerResource::all);
         get("/workers/:id", workerResource::get);
         delete("/workers/:id", workerResource::delete);
+
+        put("/experiments", experimentResource::put, Experiment.class);
+        get("/experiments", experimentResource::all);
+        get("/experiments/:id", experimentResource::get);
+        patch("/experiments/:id", experimentResource::patch, Experiment.class);
+        delete("/experiments/:id", experimentResource::delete);
+        put("/experiments/:id/answers", experimentResource::putAnswer, Experiment.class);
```
Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/Router.java]: Should be `Answer.class`.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/CalibrationResource.java [hunk 6]
```diff
@@ -42,32 +42,32 @@ public PopulationResource(PopulationOperations operations) {
...
      *
-     * @return Single population.
+     * @return a single calibration.
      */
-    public Population get(Request request, Response response) {
-        return operations.getPopulation(getParamInt(request, "id"))
-                .orElseThrow(NotFoundException::new);
+    public calibration get(Request request, Response response) {
+        return operations.getCalibration(getParamInt(request, "id"))
+                .orElseThrow(() -> new NotFoundException("Resource not found."));
```
Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/CalibrationResource.java:49]: Exception constructor should probably not be changed.

**Comment by kelunik** [comment]
Protobuf renames need a commit in https://github.com/coolcrowd/spec.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java (hunk 2 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java]: thank you! the signature is right, doIfNotRunning is already existing and i copied the method and change the logic a bit

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+        Function<ExperimentsCalibrationRecord, Calibration> toCalibration = record -> {
+            CalibrationAnswerOptionRecord a = getOrThrow(
+                    calibrationOperations.getCalibrationAnswerOption(record.getAnswer())
+            );
+            return getOrThrow(calibrationOperations.getCalibration(a.getCalibration()));
+        };
+
+        Function<Map.Entry<String, List<Calibration>>, Experiment.PlatformCalibrations> toPopulation = entry ->
+                Experiment.PlatformCalibrations.newBuilder()
+                //.setPlatformId(entry.getKey)
```
Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java]: Why is this commented out?

Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java]: good question, i think i only refactored the code, maybe it introduced a bug

Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java]: no it was commented out before - I had to wait with this until the new protobuf files where in, and forgot about it. You can comment it in. Not on keyboard right now.

Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java]: i'll do it

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java [hunk 8]
```diff
@@ -48,7 +49,31 @@ protected AbstractOperations(DSLContext create) {
                 return function.apply(trans);
             } else {
                 //TODO other exception?
```
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java]: I think this TODO is resolved ?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java [hunk 9]
```diff
@@ -48,7 +49,31 @@ protected AbstractOperations(DSLContext create) {
...
+        return create.transactionResult(trans -> {
+            boolean running = DSL.using(trans).fetchExists(
+                    DSL.selectFrom(Tables.TASK)
+                            .where(Tables.TASK.EXPERIMENT.eq(experimentID))
+                            .and(Tables.TASK.STATUS.eq(TaskStatus.running).or(Tables.TASK.STATUS.eq(TaskStatus.stopping)))
+            );
+            if (running) {
+                return function.apply(trans);
+            } else {
+                //TODO other exception?
```
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java]: I would say this exception is perfectly fine

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java (hunk 8 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java]: you are right

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java (hunk 9 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/AbstractOperations.java]: yeah i forgot to remove the TODO, thanks

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/transformers/ExperimentTransform.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+     * @param record The Database record to use
+     * @param state state of the experiment
+     * @return the experiment object with the given data
+     */
+    public static Experiment toProto(ExperimentRecord record, Experiment.State state) {
+        Type type = new TypeToken<Map<String, String>>(){}.getType();
+        return Experiment.newBuilder()
+                .setId(record.getIdExperiment())
+                .setDescription(record.getDescription())
+                        /*.set*/
```
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/transformers/ExperimentTransform.java]: TODO NEEDED ANSWER !! SELF SPANK !!

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+     * @param response Response provided by Spark.
+     * @return null on success
+     */
+    public Experiment delete(Request request, Response response) {
+        int id = getParamInt(request, "id");
+        Experiment experiment = fetchExperiment(id);
+
+        try {
+            if (!experimentOperations.deleteExperiment(id)) {
+                throw new InternalServerErrorException("Resource not found!");
```
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java]: Move to NotFoundException

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+                throw new IllegalArgumentException("Creative stop is only allowed for published experiments");
+            }
+
+            //check that there are enough datas for publish
+            if (experiment.getState().equals(Experiment.State.PUBLISHED)
+                    && experimentOperations.verifyExperimentForPublishing(id)) {
+                throw new IllegalStateException("experiment lacks information needed for publishing");
+            }
+
+            //TODO check if there is a chance to creative stopp is not possible? like no creative answers?
```
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java]: @LeanderK should we check this here ?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/Router.java [hunk 13]
```diff
@@ -99,16 +113,30 @@ public void init() {
...
 
         get("/workers/:platform/identity", workerResource::identify);
         put("/workers", workerResource::put, Worker.class);
         get("/workers", workerResource::all);
         get("/workers/:id", workerResource::get);
         delete("/workers/:id", workerResource::delete);
+
+        get("/algorithms/task_chooser", algorithmResources::allTaskChoosers);
+        get("/algorithms/answer_quality", algorithmResources::allAnswerQualityAlgorithms);
+        get("/algorithms/rating_quality", algorithmResources::allRatingQualityAlgorithms);
```
Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/Router.java]: Shouldn't these be grouped in one REST call? And better use `-` than `_` in URLs.

**Comment by LeanderK** [comment]
i think i am roughy finished. Some improvements are still possible (especially in the operations package), but that can wait! @marcelhollerbach @kelunik

**Comment by marcelhollerbach** [comment]
I am also done.

**Comment by LeanderK** [comment]
merge!

**Comment by kelunik** [comment]
\o/

**Pull request merged**